src/utils.ts

```diff
@@ -1,15 +1,54 @@
 import { SelectionEntry, UIDisplayData } from '@aics/simularium-viewer';
 import { VisibilitySelectionMap } from './constants';
 
-export const getNewSelectionMap = (
+export const getNewMapAfterTopLevelCheckboxClick = (
   agentName: string,
   currentVisibilityMap: VisibilitySelectionMap
 ): VisibilitySelectionMap => {
   const newMap: VisibilitySelectionMap = { ...currentVisibilityMap };
-  if (currentVisibilityMap[agentName].length === 0) {
+  if (currentVisibilityMap[agentName].includes(agentName)) {
+    newMap[agentName] = [];
+  } else {
     newMap[agentName] = [agentName];
+  }
+  return newMap;
+};
+
+export const getNewMapAfterDisplayStateClick = (
+  agentName: string,
+  displayStateName: string,
+  allDisplayStates: string[],
+  currentVisibilityMap: VisibilitySelectionMap
+): VisibilitySelectionMap => {
+  const newMap: VisibilitySelectionMap = { ...currentVisibilityMap };
+  const currentStates = currentVisibilityMap[agentName] || [];
+  const index = currentStates.indexOf(displayStateName);
+  if (currentStates.length === 0) {
```

Comment:
```suggestion
  if (currentStates.length === 0) {
    // Agent + all of its child display states are currently selected, so update the selected states
    // to include all of the children except for `displayStateName`.
```

Suggested comment here to explain what this is doing.

GENERAL THOUGHT: I don't love that we're trying to pack extra information into what's inside of the `currentStates` list based on whether it's empty or the agent name... It makes it much harder to read this code.

I kind of want this behavior to be controlled by an enum. Something like, `Selected.ALL`, `Selected.PARTIAL`, `Selected.NONE`, and then some payload that's included with `Selected.PARTIAL`...?

```
const enum Selected {
  PARTIAL,
  NONE,
  ALL
}

type SelectionStatus = { selected: Selected.NONE | Selected.ALL } & { selected: Selected.PARTIAL, selectedAgentChildren: string[] }
```

Reply:
Alternatively, making a human-readable method name to handle these checks would make this code easier to parse.

For example:
```
const isNoAgentChildrenSelected= (agentName: string, selectedChildren: string[]): boolean => { return selectedChildren.length === 1 && selectedChildren[0] === agentName; }

// Potentially unsafe because you could have every child in the `selectedChildren` array and it would still think that not all children are selected. 
const isAllAgentChildrenSelected = (_agentName: string, selectedChildren: string[]): boolean => { return selectedChildren.length === 0; }

const isAgentPartiallySelected = (agentName: string, selectedChildren: string[]): boolean => { return !isNoAgentChildrenSelected(agentName, selectedChildren) && !isAllAgentChildrenSelected (agentName, selectedChildren); }

if (isAllAgentChildrenSelected (agentName, currentStates)) {
 ...
} else {
  if (isNoAgentChildrenSelected(agentName, currentStates)) {
    ...
  } else {
    ...
  }
}
```

Reply:
To your general thought: I completely agree. When there is less time pressure on review I think we should have an intermediate data structure that sort of enhances/replaces `UiDisplayData` and does the work of `VisibilitySelectionMap` without the arbitrariness of the empty arrays and arrays with agent names in them.

Either way I agree that some of these changes to function names and use of enums can make this more readable and I'll start on that.